src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -21354,21 +21493,69 @@ static void F4(A4? x4, A4 y4)
                 // (32,18): warning CS8604: Possible null reference argument for parameter 'a' in 'A2.implicit operator B?(A2 a)'.
                 //         b = ((B?)x2)/*T:B?*/;
                 Diagnostic(ErrorCode.WRN_NullReferenceArgument, "x2").WithArguments("a", "A2.implicit operator B?(A2 a)").WithLocation(32, 18),
-                // (39,14): warning CS8600: Converting null literal or possible null value to non-nullable type.
-                //         b = ((B)x3)/*T:B!*/;
-                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "(B)x3").WithLocation(39, 14),
+                // (33,14): warning CS8600: Converting null literal or possible null value to non-nullable type.
+                //         b = ((B)y2)/*T:B?*/;
+                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "(B)y2").WithLocation(33, 14),
                 // (47,17): warning CS8604: Possible null reference argument for parameter 'a' in 'A4.implicit operator B(A4 a)'.
                 //         b = ((B)x4)/*T:B!*/;
                 Diagnostic(ErrorCode.WRN_NullReferenceArgument, "x4").WithArguments("a", "A4.implicit operator B(A4 a)").WithLocation(47, 17),
-                // (47,14): warning CS8600: Converting null literal or possible null value to non-nullable type.
-                //         b = ((B)x4)/*T:B!*/;
-                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "(B)x4").WithLocation(47, 14),
                 // (48,18): warning CS8604: Possible null reference argument for parameter 'a' in 'A4.implicit operator B(A4 a)'.
                 //         b = ((B?)x4)/*T:B!*/;
                 Diagnostic(ErrorCode.WRN_NullReferenceArgument, "x4").WithArguments("a", "A4.implicit operator B(A4 a)").WithLocation(48, 18));
             comp.VerifyTypes();
         }
 
+        [Fact]
+        public void ExplicitCast_UserDefined_02()
+        {
+            var source =
+@"class A1<T> where T : class
+{
+    public static implicit operator B<T?>(A1<T> a) => throw null;
+}
+class A2<T> where T : class
+{
+    public static implicit operator B<T>(A2<T> a) => throw null;
+}
+class B<T> { }
+class C<T> where T : class
+{
+    static void F1(A1<T?> x1, A1<T> y1)
+    {
+        B<T?> x;
+        B<T> y;
+        x = ((B<T?>)x1)/*T:B<T?>*/;
+        y = ((B<T>)x1)/*T:B<T!>*/;
+        x = ((B<T?>)y1)/*T:B<T?>*/;
+        y = ((B<T>)y1)/*T:B<T!>*/;
+    }
+    static void F2(A2<T?> x2, A2<T> y2)
+    {
+        B<T?> x;
+        B<T> y;
+        x = ((B<T?>)x2)/*T:B<T?>*/;
+        y = ((B<T>)x2)/*T:B<T!>*/;
+        x = ((B<T?>)y2)/*T:B<T?>*/;
+        y = ((B<T>)y2)/*T:B<T!>*/;
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (17,14): warning CS8619: Nullability of reference types in value of type 'B<T?>' doesn't match target type 'B<T>'.
```

Comment:
>// (17,14): warning CS8619: Nullability of reference types in value of type 'B<T?>' doesn't match target type 'B<T>'. [](start = 16, length = 117)

I thought we decided that explicit cast can adjust nested nullability without warnings

Reply:
I didn't think so. Getting rid of the diagnostics requires `!`.

Reply:
> I didn't think so. Getting rid of the diagnostics requires !.

@cston Could you please confirm with the latest notes. 

---
In reply to: [190754802](https://github.com/dotnet/roslyn/pull/26356#discussion_r190754802) [](ancestors = 190754802)

Reply:
I believe explicit cast should cast to the explicit type without warnings on nested nullability. I do not have a reference in the LDM notes though.

This particular issue, the nested nullability warning on `(B<T>)x1`, is a bug in `NullableWalker.ApplyConversion`. I've added comments here and in that method.

---
In reply to: [190755071](https://github.com/dotnet/roslyn/pull/26356#discussion_r190755071) [](ancestors = 190755071,190754802)